util/stmtsummary/variables.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stmtsummary
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/config"
+)
+
+const (
+	typeEnable = iota
+	typeEnableInternalQuery
+	typeRefreshInterval
+	typeHistorySize
+	typeMaxStmtCount
+	typeMaxSQLLength
+	typesNum
+)
+
+type systemVars struct {
+	sync.Mutex
+	// This array itself won't be modified once created. Only its elements may be modified.
+	variables []variable
+}
+
+type variable struct {
+	sessionValue string
+	globalValue  string
+	finalValue   int64
+}
+
+func newSysVars() *systemVars {
+	s := &systemVars{
+		variables: make([]variable, typesNum),
+	}
+	// Initialize these configurations by values in the config file.
+	// They may be overwritten by system variables later.
+	for varType := range s.variables {
+		atomic.StoreInt64(&s.variables[varType].finalValue, getConfigValue(varType))
+	}
+	return s
+}
+
+func (s *systemVars) getVariable(varType int) int64 {
+	return atomic.LoadInt64(&s.variables[varType].finalValue)
+}
+
+func (s *systemVars) setVariable(varType int, valueStr string, isSession bool) {
+	s.Lock()
+	defer s.Unlock()
+
+	v := &s.variables[varType]
+	if isSession {
+		v.sessionValue = valueStr
+	} else {
+		v.globalValue = valueStr
+	}
+	sessionValue := v.sessionValue
+	globalValue := v.globalValue
+
+	var valueInt int64
+	switch varType {
+	case typeEnable, typeEnableInternalQuery:
+		valueInt = getBoolFinalVariable(varType, sessionValue, globalValue)
+	case typeHistorySize, typeMaxSQLLength:
+		valueInt = getIntFinalVariable(varType, sessionValue, globalValue, 0)
+	case typeRefreshInterval, typeMaxStmtCount:
+		valueInt = getIntFinalVariable(varType, sessionValue, globalValue, 1)
+	default:
+		panic(fmt.Sprintf("No such type of variable: %d", varType))
+	}
+	atomic.StoreInt64(&v.finalValue, valueInt)
+}
+
+func getBoolFinalVariable(varType int, sessionValue, globalValue string) int64 {
```

Comment:
```suggestion
func getBoolFinalVariable(varType int, v variable) int64 {
```
We can change the signature of this function and we can return `finalValue` directly at L98 instead of `getConfigValue(varType)` again.

Reply:
The rule is:
1. If session variable isn't empty, use session variable.
2. If global variable isn't empty, use global variable.
3. If both are empty, use config.
So I think the code cannot be simplified.